pkg/tcpip/stack/stack.go

```diff
@@ -1161,17 +1168,32 @@ func (s *Stack) FindRoute(id tcpip.NICID, localAddr, remoteAddr tcpip.Address, n
 				continue
 			}
 
+			plen := route.Destination.Mask().Prefix()
+			if plen <= bestPrefixLen {
+				continue
+			}
+
 			if id == 0 || id == route.NIC {
 				if addressEndpoint := s.getAddressEP(nic, localAddr, remoteAddr, netProto); addressEndpoint != nil {
 					var gateway tcpip.Address
 					if needRoute {
 						gateway = route.Gateway
 					}
-					r := constructAndValidateRoute(netProto, addressEndpoint, nic /* outgoingNIC */, nic /* outgoingNIC */, gateway, localAddr, remoteAddr, s.handleLocal, multicastLoop)
-					if r == nil {
+
+					if r := constructAndValidateRoute(netProto, addressEndpoint, nic /* outgoingNIC */, nic /* outgoingNIC */, gateway, localAddr, remoteAddr, s.handleLocal, multicastLoop); r == nil {
+						panic(fmt.Sprintf("non-forwarding route validation failed with route table entry = %#v, id = %d, localAddr = %s, remoteAddr = %s", route, id, localAddr, remoteAddr))
+					} else if plen > bestPrefixLen {
+						bestRoute, bestPrefixLen = r, plen
+					}
```

Comment:
Why would we not update this before line 1171 unconditionally?

Reply:
`bestPrefixLen` starts off at 0, so at that point, *any* prefix length greater than 0 would trigger an update.

Maybe it wants to be `>=` actually, in case there's a route trying to be a default gateway route.

Reply:
I've switched it to `>=` so it will always match either on the first iteration matching the route or for a default route (prefix length 0).